File: Hard/42_trapping_rain_water/solution.py
class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        if n < 3:
            return 0

        L, R = 0, n-1
        trapped = 0
        left_max, right_max = 0, 0

        while L < R:
            left_max = max(left_max, height[L])
            right_max = max(right_max, height[R])

            if height[L] >= height[R]:
                wR = (right_max - height[R])
                trapped += wR if wR > 0 else 0
                R -= 1
            else:
                wL = (left_max - height[L])
                trapped += wL if wL > 0 else 0
                L += 1

        return trapped

File: Hard/42_trapping_rain_water/solution.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        if (n < 3) { return 0; }
        
        int trapped = 0;
        int L = 0;
        int R = n-1;
        int left_max = 0;
        int right_max = 0;

        while (L < R) {
            left_max = std::max(left_max, height[L]);
            right_max = std::max(right_max, height[R]);

            if (height[L] >= height[R]) {
                int wR = right_max - height[R];
                if (wR < 0) { wR = 0; }
                trapped += wR;
                R--;
            }
            else {
                int wL = left_max - height[L];
                if (wL < 0) { wL = 0; }
                trapped += wL;
                L++;
            }
        }

        return trapped;
    }
};
